src/main/java/com/epam/aidial/core/storage/ResourceDescription.java

```diff
@@ -146,22 +154,68 @@ public static ResourceDescription fromDecoded(ResourceDescription description, S
         return fromDecoded(description.getType(), description.getBucketName(), description.getBucketLocation(), relativePath);
     }
 
-    public static ResourceDescription fromLink(String link, EncryptionService encryptionService) {
-        String[] parts = link.split(BlobStorageUtil.PATH_SEPARATOR);
+    public static ResourceDescription fromPublicUrl(String url) {
+        return fromUrl(url, BlobStorageUtil.PUBLIC_BUCKET, BlobStorageUtil.PUBLIC_LOCATION, null);
+    }
+
+    public static ResourceDescription fromPrivateUrl(String url, ResourceDescription bucket) {
+        return fromUrl(url, bucket.getBucketName(), bucket.getBucketLocation(), null);
+    }
+
+    public static ResourceDescription fromPrivateUrl(String url, EncryptionService encryption) {
+        ResourceDescription description = fromAnyUrl(url, encryption);
+
+        if (description.isPublic()) {
+            throw new IllegalArgumentException("Not private url: " + url);
+        }
+
+        return description;
+    }
+
+    public static ResourceDescription fromAnyUrl(String url, EncryptionService encryption) {
+        return fromUrl(url, null, null, encryption);
+    }
+
+    private static ResourceDescription fromUrl(String url,
```

Comment:
can you add Nullable annotations for the arguments? It seems this method become a key point for the future usage (extension).
As far as I understand we should provide either expected bucket and location or encryptionService - not all at the same time

Reply:
You can provide all of them. Than we will decode and check against expected bucket.

Reply:
should we add a check when all 3 are null?

Reply:
It will handle public resources if all of them are null